components/script/dom/htmlinputelement.rs

```diff
@@ -1075,20 +1057,25 @@ impl HTMLInputElement {
         TextControlSelection::new(&self, &self.textinput)
     }
 
-    fn update_text_contents(&self, value: DOMString, update_text_cursor: bool) -> ErrorResult {
+    fn update_text_contents(&self, mut value: DOMString, update_text_cursor: bool) -> ErrorResult {
         match self.value_mode() {
             ValueMode::Value => {
-                // Steps 1-2.
-                let old_value = mem::replace(self.textinput.borrow_mut().single_line_content_mut(), value);
                 // Step 3.
                 self.value_dirty.set(true);
+
                 // Step 4.
                 if update_text_cursor {
-                    self.sanitize_value();
+                    self.sanitize_value(&mut value);
                 }
-                // Step 5.
-                if *self.textinput.borrow().single_line_content() != old_value {
-                    self.textinput.borrow_mut().clear_selection_to_limit(Direction::Forward, update_text_cursor);
+
+                let mut textinput = self.textinput.borrow_mut();
+
+                if *textinput.single_line_content() != value {
+                    // Steps 1-2
+                    textinput.set_content(value, update_text_cursor);
```

Comment:
Bit of a shame to reorder this, but it does prevent us setting the value, then getting it for the sanitize step and then setting it again after sanitization... and the outcome is the same.